update chart links, esp. for units/rates

huss opened the pull request:
# Description

Resource generalization added units and rates to graphics. These now work with chart links so you can reproduce these graphics with the new values. While making these changes some other items were found/addressed:

- The bug stopping map chart links from working was fixed (issue #701)
- A bug that did not start the options from the correct base URL when getting a chart link from a page that was opened with a chart link was fixed.
- The code reprocessed the chart link options on many routes until a new page URL was used. This caused a bug where bar stacking did not work and was inefficient. It was fixed.
- Fixed a bug where the bar duration was not set on the menu UI.

Fixes #701, Fixed #813

This PR is a follow on to PR #839. The initial commit on this PR reproduces that work in that PR that related to chart links. The original PR was put in by @Jumpman5 (their work, the effort of other(s) and some by @huss). Redoing the work lost the commit record from that PR but this is meant to document this. I did this because:

- The original PR had work unrelated to chart links.
- The PR was from a repo created directly on the OED organization and this caused issues in working with the PR.
- I was too lazy to fix this in another way (sorry).

## Type of change

- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

An existing issue where the slider range is not reproduced was noticed and issue #853 was created.

huss committed:
changes from PR #839

Redo changes from PR that were related to chartlinks. These were mostly
commits by @Jumpman5 but @huss did some, esp. for rate links.

huss committed:
fix & location, remove commented out code

huss committed:
minor fix for const

huss committed:
map chartlinks now work

This was an old bug.

huss committed:
fix bug where not setting bar duration in UI

huss committed:
fix bug with processing chartlinks many times

The code now only processes chartlinks once to avoid issues.

huss committed:
fix bug if create chartlink from window using chartlink

Needed to clear the parameters so don't get repeats.

huss committed:
remove debug print

huss committed:
fix linting

spearec reviewed (approved):
~~PR looks good. I tried to break it in as many possible ways, but I could not.~~
EDIT: nevermind, I spoke too soon. When changing the range slider (i.e serverRange param is not default), the chart link generated no longer loads a viable webpage.
Other than that, the links for charts using mostly default settings are needlessly large. This is obviously a purely cosmetic change.
For example, take a bar graph with just a single meter. The link given is:
`http://localhost:3000/graph?meterIDs=1&chartType=bar&serverRange=all&barDuration=28&barStacking=false&unitID=1&rate=hour,1`
However, the default values can be omitted from this link, meaning that
`http://localhost:3000/graph?meterIDs=1&chartType=bar&unitID=1`
loads the same webpage.
Would it make sense to only add parts to the link which are not default, i.e required for replicating the page?

huss commented:
I did consider the longer links but did not change them for two reasons:

1. They did not seem to cause any issues and the logic would be harder to remove others (and maybe more bugs :-)
2. The person creating the link may want to cause default behaviors when the initial graphic is changed. For example, if you have the link display a bar graphic then rate does not matter. However, if you change to a line graphic then having set the rate will modify that graphic. It probably is a matter of taste but I like it this way unless someone else has a reason not to do this.

huss committed:
fix unset state in export

The logic for lines was made similar to bars that protected against
unset state. This also caused the use of chartlinks with a defined
server range to fail due to the exception.

huss commented:
> EDIT: nevermind, I spoke too soon. When changing the range slider (i.e serverRange param is not default), the chart link generated no longer loads a viable webpage.

Good catch. It turned out that the state checks for lines did not parallel bar for export. Changing the time range caused an exception so no graph was shown. I believe my latest push fixes this.

Many thanks to @spearec for doing a careful check and code review. Let me know if you have any other thoughts.

spearec commented:
Issue seems to be fixed. PR looks good to me.
As for the default parameters in the links, maybe it's worth adding a note in the documentation about which link components are defaults are therefore optional? While I agree there's merit to always providing a full link, having a link which does not specify certain values could also be useful if an admin changes those defaults.In that case, maybe the user would want the link to match the new defaults.

spearec commented:
One more thing to note - this PR should also fix #813

huss commented:
@spearec review is being used to clear this PR so it is now ready to merge.
